### ISSUE TITLE
Authentik redirecting to Dashboard, instead of Application

### ISSUE THREAD

**Issue by AzraAnimating** [goauthentik/authentik#11980]
Hey!

I'm currently setting up Authentik in a University-Setting authenticating with Shibboleth via SAML2.

This means I'm using Authentik as the IDP inside my organization, but authenticating accounts Shibboleth using the Source Federation features. 

Currently I've set up a flow that invokes a Identification Stage where the Shibboleth is listed as the only Source. 
This means, as expected, the user is prompted to login via Shibboleth. 

The problem I'm facing is that, when the user is returned to me after authenticating with Shibboleth and the configured Authentication Flow is triggered, the user **isn't** redirected to the application they requested to be authenticated for.   
Instead, they are directed to the Dashboard which isn't accessible to them, due to their nature as an external user. 

So in short:  
I expect this redirection Path: Application -> Authentik -> Shibboleth -> Authentik -> Application  
But am getting this: Application -> Authentik -> Shibboleth -> Authentik -> Authentik Dashboard

This also only happens when the internal authorization flow is triggered, when the user is already authenticated in Authentik they are just redirected to the application as expected.

When testing this with other Source-Types (Oauth etc.) i've never encountered this problem. But I'm unsure if this is a bug or something related to the way SAML2 works. 

Thanks for reading! Any and all help would be much appreciated, as I'm very much out of ideas.

**Version and Deployment**

-   authentik version: 2024.8.4
-   Deployment: Kubernetes

**Comment by D-Sekiryuutei** [comment]
This sounds like a configuration issue with how Authentik handles
redirection after authentication via SAML2 with Shibboleth. Here are some
possible directions to troubleshoot:

1. Redirect URI/Target Application Configuration: It seems that the
application redirection path after Shibboleth authentication isn’t mapped
back to the original application. Check if there’s a specific configuration
within Authentik's SAML setup for Shibboleth that specifies the
post-authentication redirect URI. Ensure it’s pointing to the original
application URL rather than the Authentik dashboard.

2. Check Authentik Authentication Flow: Since other source types (OAuth)
work fine, the issue might be specific to the SAML flow configuration.
Authentik’s flow editor might allow you to specify a "default" or
"post-authentication" redirection URI that should lead back to the
application, not the dashboard. Try editing the flow to add an “Application
Redirect” stage or explicitly configure a redirect override in the SAML2
identification stage.

3. Mapping SAML2 Relay State: SAML often uses a RelayState parameter to
maintain the original request URL and redirect back after authentication.
Ensure that the RelayState is correctly maintained through the flow. If
this is missing or overwritten during the process, it might default to the
dashboard.

4. Inspect Logs and SAML Traces: Examine the logs for any hints about how
Authentik is handling the callback after Shibboleth authentication.
Sometimes, Authentik’s logs or debug mode can reveal issues in how the
redirection is managed.

5. Look for Known Issues: Authentik's GitHub issues page or forum might
have users with similar issues. Given that this only happens with SAML2,
there could be a known limitation or bug with the current version that
affects the handling of RelayState or redirection post-authentication.

If these steps don’t resolve it, sharing logs and more details with the
Authentik community could provide insights since it may be an edge case
with SAML2 handling in the Kubernetes deployment.

**Comment by AzraAnimating** [comment]
Okay: After a lot of trial and error we've found a somewhat satisfactory solution. 

First of all: we've tracked down, that the redirect-url gets set incorrectly or not at all, resulting in the redirect to the Dashboard.

We've set an Expression Policy for the Identification-Stage containing the SAML2 Source. 
This policy correctly sets the redirect attribute, so it can't  be lost. 

Specifically our solution looks like this: 

```python
plan = request.context["flow_plan"]

# Get the request query
query_string = request.http_request.GET.get("query")
start_index = query_string.find("next=")

# If the next parameter is set in the query string
if start_index != -1:
    next_param = query_string[start_index + len("next="):]

    # Redirect to the SAML auth with the given next value
    if next_param:
      # Take the encoded query string and add it to the next parameter
      plan.redirect("/source/saml/<source-name>/?next=" + next_param)
      return False

# Otherwise fail
return False
```

(Maybe this is an authentik-issue, but I'll investigate this further later, for now I'm just happy, that it works.)

I hope this helps someone!

**Comment by phmcder** [comment]
I am seeing you the same issue whilst authenticating with SAML. Is there a decision if this is a bug or not?

**Comment by AzraAnimating** [comment]
I'd think so, haven't gotten around to validating / reporting that yet tho...

**Comment by ZandercraftGames** [comment]
Sorry to necro on this issue, but I've been getting this since a more recent update with Microsoft as an auth source.

**Comment by AzraAnimating** [comment]
Well. After upgrading to 2025.2.3 this behavior is back.
Re-Introduced by #13696 and currently discussed in #13719
